vopr: fix false positive

matklad opened the pull request:
When looking for available replies, we should only consider .up replicas, otherwise we are stepping into uninitialized memory!

Seed: ./zig/zig build -Drelease vopr -Dvopr-state-machine=testing -- 11119430380926947877

matklad commented on src/vopr.zig:
```diff
@@ -1146,7 +1146,8 @@ pub const Simulator = struct {
         var replies_latest = [_]?vsr.Header.Reply{null} ** constants.clients_max;
-        for (simulator.cluster.replicas) |replica| {
+        for (simulator.cluster.replicas, simulator.cluster.replica_health) |replica, health| {
+            if (health == .down) continue;
```
TBH, something feels off about this check:

* for any cluster run, imagine a run where every replica not in core is down
* then this loop is equivalent to a loop over only core replicas
* but then it tautologically returns true

Should we perhaps track required replies in some other way? Should we perhaps extend StateChecker's `on_message` to track replies and explicitly capture which replies we _should_ have?

chaitanyabhandari replied:
Mhmmmmmm interesting case indeed! I was thinking that we could *also* solve this by not introducing additional logic in `on_message`, and instead just making sure the memory for replicas that are down is valid? 

I had done that in https://github.com/tigerbeetle/tigerbeetle/pull/2708; instead of calling `deinit` in `replica_crash` like we currently do, we call `replica_deinit` _before_ every invocation to `replica_open` (and also in `deinit` in cluster.zig). https://github.com/tigerbeetle/tigerbeetle/blob/74847f2f8fd1bb457fd66282562ee55e0c3fd7ff/src/testing/cluster.zig#L557-L569

Thoughts @matklad?

matklad replied:
Oh, I like that! Having the memory available would also be super helpful for debugging the issues, and is worth doing just for that reason!

Though, separately from this, I guess I am unsure about the approach we have right now, where we trust _some_ replica's memory to retain the right answer. So, if replica's collude to erase their memory (due to a bug), we wouldn't notice it! I'd be more comfortable checking against _externalized_ state (messages sent), but that's a separate thing.

matklad replied:
Ah, this doesn't quite work! I was hoping we can achieve an invariant that all cluster's replicas memories are always valid, but this doesn't actually work. A replica can fail to open, in which case we won't have any valid memory at all to keep around.

This happens in "Cluster: recovery: WAL prepare corruption (R=1, corrupt between checkpoint and head)"

matklad replied:
Ok, take three in!

chaitanyabhandari replied:
Yes, the invariant that all the replicas' memories are always valid doesn't actually work, for the case of wild VOPR even more so because `open` can fail there during VOPR runs as well, due to `WALCorrupt`! I had to introduce a new status called `dead` there, and ensure that we don't have `dead` replicas to `simulator.core`.

Also, indeed the case where the replica memory is untrustworthy due to a bug is a tricky one, I didn't understand that that was what you were referring to before! I'm going to review take 3 now.

matklad committed:
vopr: fix false positive

In core_missing_reply, we find the reply the core _should_ have
iterating over all replicas, and we can stepe into uninitialized memory
if a replica is down.

To fix it, compute the set of replies we _should_ have without
referening replicas local memories, via the state checker.

This feels more proper --- rather than trusting replicas about their own
state, we verify state against anything that was externalized.

Seed: ./zig/zig build -Drelease vopr -Dvopr-state-machine=testing -- 11119430380926947877

chaitanyabhandari commented on src/testing/cluster/state_checker.zig:
```diff
@@ -72,25 +80,66 @@ pub fn StateCheckerType(comptime Client: type, comptime Replica: type) type {
+
+        pub fn on_client_eviction(state_checker: *StateChecker, client_id: u128) void {
+            const removed = state_checker.client_replies.swapRemove(client_id);
+            maybe(removed);
```
Mhmmmmm this is subtle! The way I understood it was that even when a client has been evicted from `replica.client_sessions` and removed from `state_checker.client_replies`, a replayed `register` message can cause it to get re-added to `replica.client_sessions`. However, it may not get added to `state_checker.client_replies` again due to the primary crashing (and no reply message being sent). Then, if the client is evicted from `replica.client_sessions` again, this `swamRemove` would return `False`!

Does that sound about right @matklad or is there a more straightforward case I'm missing?

matklad replied:
I was thinking of a simpler case: Primary A adds a client to its client_sessions, evicting someone, and then crashes. Primary B takes over, and adds replays the eviction.

chaitanyabhandari replied:
Ah yes, thank you!

chaitanyabhandari reviewed (approved):
LGTM; I like the approach of tracking new clients using `clients_register_op_latest`. This is also VOPR green at ~350k seeds, approving!
